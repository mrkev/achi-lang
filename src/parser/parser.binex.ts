import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { Node, withAt } from "./Node";

type PrefixUnaryOperator = "-" | "!";
export type PrefixUnaryOperation = Node<{
  kind: "PrefixUnaryOperation";
  operator: PrefixUnaryOperator;
  value: NEXT_PARSER;
}>;

type SuffixUnaryOperator = "!";
export type SuffixUnaryOperation = Node<{
  kind: "SuffixUnaryOperation";
  operator: SuffixUnaryOperator;
  value: NEXT_PARSER;
}>;

type BinaryOperator =
  // String + Numerial
  | "+"
  // Numerical
  | "-"
  | "*"
  | "/"
  | "^"
  // Boolean
  | "&&"
  | "||"
  // Comparison
  | ">"
  | "<"
  | ">="
  | "<="
  | "=="
  | "!=";

export type BinaryOperation = Node<{
  kind: "BinaryOperation";
  operator: BinaryOperator;
  left: NEXT_PARSER;
  right: NEXT_PARSER;
}>;

export type OperatableExpression =
  | LangType["NumberLiteral"]
  | LangType["ValueIdentifier"]
  | LangType["StringLiteral"]
  | LangType["BooleanLiteral"]
  | LangType["FunctionCall"]
  | LangType["NullLiteral"];

// This parser supports basic math with + - * / ^, unary negation, factorial,
// and parentheses. It does not evaluate the math, just turn it into a series of
// nested lists that are easy to evaluate.

// You might think that parsing math would be easy since people learn it early
// in school, but dealing with precedence and associativity of operators is
// actually one of the hardest and most tedious things you can do in a parser!
// If you look at a language like JavaScript, it has even more operators than
// math, like = and && and || and ++ and so many more...

///////////////////////////////////////////////////////////////////////

type NEXT_PARSER =
  | OperatableExpression
  | PrefixUnaryOperation
  | SuffixUnaryOperation
  | BinaryOperation;

// Takes a parser for the prefix operator, and a parser for the base thing being
// parsed, and parses as many occurrences as possible of the prefix operator.
// Note that the parser is created using `Parsimmon.lazy` because it's recursive. It's
// valid for there to be zero occurrences of the prefix operator.
function PREFIX(
  operatorsParser: Parsimmon.Parser<OpLoc<PrefixUnaryOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
): Parsimmon.Parser<NEXT_PARSER> {
  const parser = Parsimmon.lazy<NEXT_PARSER>(() => {
    const res: Parsimmon.Parser<NEXT_PARSER> = Parsimmon.seqMap(
      Parsimmon.index,
      operatorsParser,
      parser,
      Parsimmon.index,
      (start, operator, value, end) =>
        ({
          kind: "PrefixUnaryOperation",
          operator: operator.op,
          value,
          "@": { start, end },
        } as const)
    ).or(nextParser);
    return res;
  });
  return parser;
}

type OpLoc<T extends string> = {
  op: T;
  start: Parsimmon.Index;
  end: Parsimmon.Index;
};

// Ideally this function would be just like `PREFIX` but reordered like
// `Parsimmon.seq(parser, operatorsParser).or(nextParser)`, but that doesn't work. The
// reason for that is that Parsimmon will get stuck in infinite recursion, since
// the very first rule. Inside `parser` is to match parser again. Alternatively,
// you might think to try `nextParser.or(Parsimmon.seq(parser, operatorsParser))`, but
// that won't work either because in a call to `.or` (aka `Parsimmon.alt`), Parsimmon
// takes the first possible match, even if subsequent matches are longer, so the
// parser will never actually look far enough ahead to see the postfix
// operators.
function POSTFIX(
  operatorsParser: Parsimmon.Parser<OpLoc<"!">>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
) {
  // Because we can't use recursion like stated above, we just match a flat list
  // of as many occurrences of the postfix operator as possible, then use
  // `.reduce` to manually nest the list.
  //
  // Example:
  //
  // INPUT  :: "4!!!"
  // PARSE  :: [4, "factorial", "factorial", "factorial"]
  // REDUCE :: ["factorial", ["factorial", ["factorial", 4]]]
  return Parsimmon.seqMap(nextParser, operatorsParser.many(), (x, suffixes) =>
    suffixes.reduce((acc, { op, end }) => {
      return {
        kind: "SuffixUnaryOperation",
        operator: op,
        value: acc,
        "@": {
          start: acc["@"].start,
          end,
        },
      } as const;
    }, x)
  );
}

// Takes a parser for all the operators at this precedence level, and a parser
// that parsers everything at the next precedence level, and returns a parser
// that parses as many binary operations as possible, associating them to the
// right. (e.g. 1^2^3 is 1^(2^3) not (1^2)^3)
function BINARY_RIGHT(
  operatorsParser: Parsimmon.Parser<OpLoc<BinaryOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
): Parsimmon.Parser<NEXT_PARSER> {
  const parser = Parsimmon.lazy<NEXT_PARSER>(() => {
    const chainRes: Parsimmon.Parser<NEXT_PARSER> = nextParser.chain((next) =>
      withAt(
        Parsimmon.seqMap(
          operatorsParser,
          Parsimmon.of(next),
          parser,
          (operator, next, p) =>
            ({
              kind: "BinaryOperation",
              operator: operator.op,
              left: next,
              right: p,
            } as const)
        )
      ).or(Parsimmon.of(next))
    );

    return chainRes;
  });
  return parser;
}

// Takes a parser for all the operators at this precedence level, and a parser
// that parsers everything at the next precedence level, and returns a parser
// that parses as many binary operations as possible, associating them to the
// left. (e.g. 1-2-3 is (1-2)-3 not 1-(2-3))
function BINARY_LEFT(
  operatorsParser: Parsimmon.Parser<OpLoc<BinaryOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
) {
  // We run into a similar problem as with the `POSTFIX` parser above where we
  // can't recurse in the direction we want, so we have to resort to parsing an
  // entire list of operator chunks and then using `.reduce` to manually nest
  // them again.
  //
  // Example:
  //
  // INPUT  :: "1+2+3"
  // PARSE  :: [1, ["+", 2], ["+", 3]]
  // REDUCE :: ["+", ["+", 1, 2], 3]
  return Parsimmon.seqMap(
    nextParser,
    Parsimmon.seq(operatorsParser, nextParser).many(),
    (first, rest) => {
      return rest.reduce((acc, ch) => {
        const [op, another] = ch;
        return {
          kind: "BinaryOperation",
          operator: op.op,
          left: acc,
          right: another,
          "@": {
            start: acc["@"].start,
            end: another["@"].end,
          },
        } as const;
      }, first);
    }
  );
}

export function OperatorParser(r: Parsimmon.TypedLanguage<LangType>) {
  // Operators should allow whitespace around them, but not require it. This
  // helper combines multiple operators together with names.
  //
  // Example: operators({Add: "+", Sub: "-"})
  //
  // Gives back an operator that parses either + or - surrounded by optional
  // whitespace, and gives back the word "Add" or "Sub" instead of the character.
  function operators<Ops extends readonly string[]>(
    ops: Ops
  ): Parsimmon.Parser<OpLoc<Ops[number]>> {
    const withWhitespace = ops.map((sym) =>
      Parsimmon.seqMap(
        Parsimmon.index,
        r._,
        Parsimmon.string(sym),
        r._,
        Parsimmon.index,
        (start, _0, op, _2, end) => {
          return { op, start, end };
        }
      )
    );
    const res = Parsimmon.alt(...withWhitespace);
    return res;
  }

  // A basic value is any parenthesized expression or a number.
  // TODO: start, end don't take the parens into consideration. Is that an issue?
  const Basic: Parsimmon.Parser<NEXT_PARSER> = Parsimmon.lazy<NEXT_PARSER>(() =>
    Parsimmon.string("(")
      .then(MyMath)
      .skip(Parsimmon.string(")"))
      .or(r.NumberLiteral)
      .or(r.StringLiteral)
      .or(r.BooleanLiteral)
      .or(r.FunctionCall)
      .or(r.NullLiteral)
      .or(r.ValueIdentifier)
  );

  // Now we can describe the operators in order by precedence. You just need to
  // re-order the table.
  // const table = [
  //   { type: POSTFIX, ops: operators(["!"] as const) },
  //   { type: PREFIX, ops: operators(["-", "!"] as const) },
  //   { type: BINARY_RIGHT, ops: operators(["^"] as const) },
  //   {
  //     type: BINARY_LEFT,
  //     ops: operators(["*", "/"] as const),
  //   },
  //   {
  //     type: BINARY_LEFT,
  //     ops: operators(["+", "-"] as const),
  //   },
  // ];

  const table: Array<
    (next: Parsimmon.Parser<NEXT_PARSER>) => Parsimmon.Parser<NEXT_PARSER>
  > = [
    (next) => BINARY_LEFT(operators(["==", "!="] as const), next),
    (next) => POSTFIX(operators(["!"] as const), next),
    (next) => PREFIX(operators(["-", "!"] as const), next),
    (next) => BINARY_RIGHT(operators(["^"] as const), next),
    (next) => BINARY_LEFT(operators(["*", "/"] as const), next),
    (next) => BINARY_LEFT(operators(["+", "-"] as const), next),
    (next) => BINARY_LEFT(operators(["&&"] as const), next),
    (next) => BINARY_LEFT(operators(["||"] as const), next),
    (next) => BINARY_LEFT(operators([">=", "<=", ">", "<"] as const), next),
  ];

  // Start off with Num as the base parser for numbers and thread that through the
  // entire table of operator parsers.
  const tableParser = table.reduce((next, level) => level(next), Basic);

  // The above is equivalent to:
  //
  // TYPE(operators({...}),
  //   TYPE(operators({...}),
  //     TYPE(operators({...})),
  //       TYPE(operators({...}),
  //         TYPE(operators({...}), ...))))
  //
  // But it's easier if to see what's going on and reorder the precedence if we
  // keep it in a table instead of nesting it all manually.

  // This is our version of a math expression.
  const MyMath = tableParser.trim(r._);
  return MyMath;
}
