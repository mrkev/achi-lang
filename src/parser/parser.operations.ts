import { LangType } from "./parser";
import {
  BinaryOperation,
  OperatableExpression,
  OperatorParser,
  PrefixUnaryOperation,
  SuffixUnaryOperation,
} from "./parser.binex";
import { sublang } from "./sublang";

export type LangType_BinOp = {
  OperationExpression:
    | OperatableExpression
    | PrefixUnaryOperation
    | SuffixUnaryOperation
    | BinaryOperation;
};

/**
 * Defines unary and binary operations
 */
export const LangDef_BinOp = sublang<LangType, LangType_BinOp>({
  OperationExpression: OperatorParser,
});
