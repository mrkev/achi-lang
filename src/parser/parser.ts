import * as Parsimmon from "parsimmon";
import { LangType_BinOp, LangDef_BinOp } from "./parser.binop";
import { LangType_Match, LangDef_Match } from "./parser.match";

type LT_NonNodeKeys = {
  [Key in keyof LangType]: LangType[Key] extends string | number | symbol
    ? Key
    : never;
}[keyof LangType];

// LangType, but skipping any entry that isn't an object
type LT_OnlyNodes = Omit<LangType, LT_NonNodeKeys>;
type ValueOf<T> = T[keyof T];
// example:
//   ExhaustiveParsers<LangType["Statement"]>
// can only be satisfied by an object that entries for all the types
// that make up LangType["Statement"]. Each entry maps to a Parsimmon.Parser.
type ExhaustiveParsers<T extends ValueOf<LT_OnlyNodes>> = {
  [key in T["kind"]]: Parsimmon.Parser<LangType[key]>;
};

// From: https://github.com/jneen/parsimmon/blob/master/examples/json.js

// Use the JSON standard's definition of whitespace rather than Parsimmon's.
let whitespace = Parsimmon.regexp(/\s*/m);

// JSON is pretty relaxed about whitespace, so let's make it easy to ignore
// after most text.
function token<T>(parser: Parsimmon.Parser<T>) {
  return parser.skip(whitespace);
}

// Turn escaped characters into real ones (e.g. "\\n" becomes "\n").
function interpretEscapes(str: string) {
  const escapes = {
    b: "\b",
    f: "\f",
    n: "\n",
    r: "\r",
    t: "\t",
  };
  return str.replace(/\\(u[0-9a-fA-F]{4}|[^u])/, (_, escape: string) => {
    let type = escape.charAt(0);
    let hex = escape.slice(1);
    if (type === "u") {
      return String.fromCharCode(parseInt(hex, 16));
    }
    if (escapes.hasOwnProperty(type)) {
      return (escapes as any)[type];
    }
    return type;
  });
}

export type LangType = LangType_BinOp &
  LangType_Match & {
    _: string;
    __: string;
    _comma: string;
    _nl: string;
    __nl: string;
    blockComment: string;

    // Expressions
    ValueIdentifier: { kind: "ValueIdentifier"; value: string };
    TypeIdentifier: { kind: "TypeIdentifier"; value: string };
    NumberLiteral: { kind: "NumberLiteral"; value: number };
    StringLiteral: { kind: "StringLiteral"; value: string };
    BooleanLiteral: { kind: "BooleanLiteral"; value: boolean };

    NestedTypeIdentifier: {
      kind: "NestedTypeIdentifier";
      path: Array<LangType["TypeIdentifier"]>;
    };

    ReturnStatement: {
      kind: "ReturnStatement";
      expression: LangType["Expression"];
    };

    IfStatement: {
      kind: "IfStatement";
      guard: LangType["Expression"];
      block: LangType["Block"];
    };

    Expression:
      | LangType["BooleanLiteral"]
      | LangType["FunctionDefinition"]
      | LangType["MatchExpression"]
      | LangType["NumberLiteral"]
      | LangType["NamedRecordLiteral"]
      | LangType["FunctionCall"]
      | LangType["ValueIdentifier"]
      | LangType["RecordLiteral"]
      | LangType["StringLiteral"];

    FunctionCall: {
      kind: "FunctionCall";
      identifier: LangType["ValueIdentifier"];
      argument: LangType["RecordLiteral"];
    };

    FunctionDefinition: {
      kind: "FunctionDefinition";
      argument: LangType["RecordDefinition"];
      body: LangType["Block"];
    };

    ConstantDefinition: {
      kind: "ConstantDefinition";
      identifier: LangType["ValueIdentifier"];
      expression: LangType["Expression"];
    };

    // Records

    // Point(x: number, y: number)
    NamedRecordDefinition: {
      kind: "NamedRecordDefinition";
      identifier: LangType["TypeIdentifier"];
      record: LangType["RecordDefinition"];
    };

    // class Point(x: number, y: number)
    NamedRecordDefinitionStatement: {
      kind: "NamedRecordDefinitionStatement";
      namedRecordDefinition: LangType["NamedRecordDefinition"];
    };

    // classes Card { King(); Queen(); Jack(); Number(value: number) }
    NamedRecordDefinitionGroup: {
      kind: "NamedRecordDefinitionGroup";
      identifier: LangType["TypeIdentifier"];
      namedRecordDefinitions: Array<LangType["NamedRecordDefinition"]>;
    };

    // Point(x: 5, y: 3)
    NamedRecordLiteral: {
      kind: "NamedRecordLiteral";
      identifier: LangType["TypeIdentifier"] | LangType["NestedTypeIdentifier"];
      recordLiteral: LangType["RecordLiteral"];
    };

    //* (x: 5, y: 5)
    RecordLiteral: {
      kind: "RecordLiteral";
      definitions: Array<LangType["NamedLiteral"]>;
    };

    //* (x: number, y: number)
    RecordDefinition: {
      kind: "RecordDefinition";
      definitions: Array<LangType["NamedDefinition"]>;
    };

    // x: 5
    NamedLiteral: {
      kind: "NamedLiteral";
      identifier: LangType["ValueIdentifier"];
      expression: LangType["Expression"];
    };

    // x: number
    NamedDefinition: {
      kind: "NamedDefinition";
      identifier: LangType["ValueIdentifier"];
      typeTag: LangType["TypeTag"];
    };

    TypeExpression:
      | LangType["NamedRecordDefinition"]
      | LangType["RecordDefinition"]
      | LangType["TypeIdentifier"];

    // :string
    TypeTag: {
      kind: "TypeTag";
      identifier: LangType["TypeIdentifier"];
    };

    // #log value
    DEBUG_Log: {
      kind: "DEBUG_Log";
      expression: LangType["Expression"];
    };

    // #log Type
    DEBUG_LogType: {
      kind: "DEBUG_LogType";
      typeExpression: LangType["TypeExpression"];
    };

    Statement:
      | LangType["ReturnStatement"]
      | LangType["IfStatement"]
      | LangType["MatchExpression"]
      | LangType["NamedRecordDefinitionStatement"]
      | LangType["NamedRecordDefinitionGroup"]
      | LangType["ConstantDefinition"]
      | LangType["MatchFunction"]
      | LangType["FunctionCall"]
      | LangType["DEBUG_LogType"]
      | LangType["DEBUG_Log"];

    Program: { kind: "Program"; statements: Array<LangType["Statement"]> };
    StatementList: {
      kind: "StatementList";
      statements: Array<LangType["Statement"]>;
    };
    Block: { kind: "Block"; statements: Array<LangType["Statement"]> };

    List: any;
    TupleDefinition: any;
    NamedTupleDefinition: any;
  };

export const Lang = Parsimmon.createLanguage<LangType>({
  _: (r) => {
    return Parsimmon.alt(r.blockComment, Parsimmon.optWhitespace);
  },
  __: () => {
    return Parsimmon.whitespace;
  },
  _comma: (r) => {
    return Parsimmon.string(",").trim(r._);
  },
  _nl: (r) => {
    return Parsimmon.regex(/[\s;]*/).trim(r._);
  },
  __nl: (r) => {
    return Parsimmon.alt(Parsimmon.string(";"), Parsimmon.string("\n")).trim(
      r._
    );
  },

  // /* ... */
  blockComment: (r) => {
    return Parsimmon.regex(/\/\*[\s\S]*\*\//);
  },

  ////////////////////////////////////////////////////////////// Expressions ///

  NumberLiteral: () => {
    return Parsimmon.regexp(/[0-9]+/).map((v) => ({
      kind: "NumberLiteral",
      value: Number(v),
    }));
  },

  // Regexp based parsers should generally be named for better error reporting.
  StringLiteral: () =>
    token(Parsimmon.regexp(/"((?:\\.|.)*?)"/, 1))
      .map(interpretEscapes)
      .map(
        (str) =>
          ({
            kind: "StringLiteral",
            value: str,
          } as const)
      )
      .desc("string"),

  BooleanLiteral: () => {
    return Parsimmon.regexp(/true|false/).map((str) => ({
      kind: "BooleanLiteral",
      value: str === "true",
    }));
  },

  ...LangDef_BinOp,

  // Card, Point, SomeDataStructure, number
  TypeIdentifier: () => {
    return Parsimmon.regexp(
      /string|number|boolean|object|array|[A-Z][a-zA-Z]*/
    ).map((v) => ({
      kind: "TypeIdentifier",
      value: v,
    }));
  },

  // TODO: test
  NestedTypeIdentifier: (r) => {
    return Parsimmon.sepBy1(r.TypeIdentifier, Parsimmon.string(".")).map(
      (path) => {
        return {
          kind: "NestedTypeIdentifier",
          path,
        };
      }
    );
  },

  // card, point, doThisOrThat
  ValueIdentifier: () => {
    return Parsimmon.regexp(
      /(?!string|number|boolean|object|array)[a-z][a-zA-Z]*/
    ).map((v) => ({
      kind: "ValueIdentifier",
      value: v,
    }));
  },

  // TODO: remove?
  List: (r) => {
    return Parsimmon.string("[")
      .then(r.Expression.sepBy(r._))
      .skip(Parsimmon.string("]"));
  },

  Expression: (r) => {
    const expressionParsers: ExhaustiveParsers<LangType["Expression"]> = {
      BooleanLiteral: r.BooleanLiteral,
      FunctionDefinition: r.FunctionDefinition,
      MatchExpression: r.MatchExpression,
      NumberLiteral: r.NumberLiteral,
      NamedRecordLiteral: r.NamedRecordLiteral,
      FunctionCall: r.FunctionCall,
      ValueIdentifier: r.ValueIdentifier,
      RecordLiteral: r.RecordLiteral,
      StringLiteral: r.StringLiteral,
    };
    return Parsimmon.alt<LangType["Expression"]>(
      ...Object.values(expressionParsers)
    );
  },

  ////////

  Program: (r) => {
    return r.StatementList.map(({ statements }) => {
      return {
        kind: "Program",
        statements,
      };
    });
  },

  // return 5; f(); class Point()
  StatementList: (r) => {
    return Parsimmon.sepBy(r.Statement, r.__nl)
      .trim(r._nl)
      .map((statements) => {
        return {
          kind: "StatementList",
          statements,
        };
      });
  },

  Statement: (r) => {
    const statementParsers: ExhaustiveParsers<LangType["Statement"]> = {
      ReturnStatement: r.ReturnStatement,
      IfStatement: r.IfStatement,
      MatchFunction: r.MatchFunction,
      MatchExpression: r.MatchExpression,
      NamedRecordDefinitionStatement: r.NamedRecordDefinitionStatement,
      NamedRecordDefinitionGroup: r.NamedRecordDefinitionGroup,
      ConstantDefinition: r.ConstantDefinition,
      FunctionCall: r.FunctionCall,
      DEBUG_LogType: r.DEBUG_LogType,
      DEBUG_Log: r.DEBUG_Log,
    };
    return Parsimmon.alt<LangType["Statement"]>(
      ...Object.values(statementParsers)
    );
  },

  ...LangDef_Match,

  // return 5
  ReturnStatement: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("return"),
      r.__,
      r.Expression,
      (_0, _1, expression) => {
        return {
          kind: "ReturnStatement",
          expression,
        };
      }
    );
  },

  // NOTE: maybe support no parentheses too
  IfStatement: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("if"),
      r._,
      Parsimmon.string("("),
      r._,
      r.Expression,
      r._,
      Parsimmon.string(")"),
      r._,
      r.Block,
      (_0, _1, _2, _3, expression, _5, _6, _7, block) => {
        return {
          kind: "IfStatement",
          guard: expression,
          block,
        };
      }
    );
  },

  // #log <expression> logs for debugging
  // we don't have function calls yet but this will do
  DEBUG_Log: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("#log"),
      r.__,
      r.Expression,
      function (_0, _1, expression) {
        return {
          kind: "DEBUG_Log",
          expression,
        };
      }
    );
  },

  DEBUG_LogType: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("#logtype"),
      r.__,
      r.TypeExpression,
      function (_0, _1, typeExpression) {
        return {
          kind: "DEBUG_LogType",
          typeExpression,
        };
      }
    );
  },

  //////////////

  // const x = 3
  ConstantDefinition: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("const"),
      r.__,
      r.ValueIdentifier,
      // TODO
      // r.TypeTag,
      r._,
      Parsimmon.string("="),
      r._,
      r.Expression,
      function (_0, _1, identifier, _2, _3, _4, expression) {
        return {
          kind: "ConstantDefinition",
          identifier,
          expression,
        };
      }
    );
  },

  /* 
  TODO;
  const add = (x: number, y: number) => number {
  }

  function foo(x: number): number {
    return x + 3;
  }
   */
  FunctionDefinition: (r) => {
    return Parsimmon.seqMap(
      r.RecordDefinition,
      r._,
      Parsimmon.string("=>"),
      // TODO: type tag?
      r._,
      r.Block,
      function (argument, _1, _2, _3, body) {
        return {
          kind: "FunctionDefinition",
          argument,
          body,
        };
      }
    );
  },

  Block: (r) => {
    return (
      Parsimmon.string("{")
        // TBD: sepBy1 and make a different parser for Unit?
        .then(r.StatementList)
        .skip(Parsimmon.string("}"))
        .map(({ statements }) => {
          return {
            kind: "Block",
            statements,
          };
        })
    );
  },

  // log(msg: "hello")
  FunctionCall: (r) => {
    return Parsimmon.seqMap(
      r.ValueIdentifier,
      r.RecordLiteral,
      function (identifier, argument) {
        return {
          kind: "FunctionCall",
          identifier,
          argument,
        };
      }
    );
  },

  // type Point(number, number)
  NamedTupleDefinition: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("type"),
      r.__,
      r.TypeIdentifier,
      r._,
      r.TupleDefinition,
      function (_0, _1, identifier, _2, tuple) {
        return {
          kind: "NamedTupleDefinition",
          identifier,
          tuple,
        };
      }
    );
  },

  //* (number, number)
  TupleDefinition: (r) => {
    return Parsimmon.string("(")
      .then(Parsimmon.sepBy1(r.TypeIdentifier, r._comma))
      .skip(Parsimmon.string(")"))
      .map((definitions) => {
        return {
          kind: "TupleDefinition",
          definitions,
        };
      });
  },

  ////////////////////////////////////////////////////////////////// Records ///

  //* Point(x: 5, y: 5)
  NamedRecordLiteral: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.alt<
        LangType["TypeIdentifier"] | LangType["NestedTypeIdentifier"]
      >(
        r.TypeIdentifier.notFollowedBy(Parsimmon.string(".")),
        r.NestedTypeIdentifier
      ),
      r.RecordLiteral,
      function (identifier, recordLiteral) {
        return {
          kind: "NamedRecordLiteral",
          identifier,
          recordLiteral,
        };
      }
    );
  },

  //* Point(x: number, y: number)
  NamedRecordDefinition: (r) => {
    return Parsimmon.seqMap(
      r.TypeIdentifier,
      r.RecordDefinition,
      function (identifier, record) {
        return {
          kind: "NamedRecordDefinition",
          identifier,
          record,
        };
      }
    );
  },

  //* class Point(x: number, y: number)
  NamedRecordDefinitionStatement: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("class"),
      r._,
      r.NamedRecordDefinition,
      function (_0, _1, namedRecordDefinition) {
        return {
          kind: "NamedRecordDefinitionStatement",
          namedRecordDefinition,
        };
      }
    );
  },

  // classes Card { King(); Queen(); Jack(); Number(value: number); }
  NamedRecordDefinitionGroup: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("classes"),
      r._,
      r.TypeIdentifier,
      r._,
      Parsimmon.string("{"),
      r._,
      Parsimmon.sepBy1(r.NamedRecordDefinition, r.__nl),
      r._,
      Parsimmon.string("}"),
      function (_0, _1, identifier, _3, _4, _5, nrdList, _7, _8) {
        return {
          kind: "NamedRecordDefinitionGroup",
          identifier,
          namedRecordDefinitions: nrdList,
        };
      }
    );
  },

  //* (x: 5, y: 5)
  RecordLiteral: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("("),
      r._,
      Parsimmon.sepBy(r.NamedLiteral, r._comma),
      r._,
      Parsimmon.string(")"),
      (_0, _1, definitions, _3, _4) => {
        return {
          kind: "RecordLiteral",
          definitions,
        };
      }
    );
  },

  //* (x: number, y: number)
  RecordDefinition: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("("),
      r._,
      // TBD: sepBy1 and make a different parser for Unit?
      Parsimmon.sepBy(r.NamedDefinition, r._comma),
      r._,
      Parsimmon.string(")"),
      function (_0, _1, definitions, _2, _3) {
        return {
          kind: "RecordDefinition",
          definitions,
        };
      }
    );
  },

  /////////////////////////////////////////////////////// Naming expressions ///

  // x: 5
  NamedLiteral: (r) => {
    return Parsimmon.seqMap(
      r.ValueIdentifier,
      r._,
      Parsimmon.string(":"),
      r._,
      r.Expression,
      (identifier, _1, _2, _3, expression) => {
        return {
          kind: "NamedLiteral",
          identifier,
          expression,
        };
      }
    );
  },

  // x: number
  NamedDefinition: (r) => {
    return Parsimmon.seqMap(
      r.ValueIdentifier,
      r._,
      r.TypeTag,
      (identifier, _3, typeTag) => {
        return {
          kind: "NamedDefinition",
          identifier,
          typeTag,
        };
      }
    );
  },

  TypeExpression: (r) => {
    const typeParsers: ExhaustiveParsers<LangType["TypeExpression"]> = {
      TypeIdentifier: r.TypeIdentifier,
      RecordDefinition: r.RecordDefinition,
      NamedRecordDefinition: r.NamedRecordDefinition,
    };
    return Parsimmon.alt<LangType["TypeExpression"]>(
      ...Object.values(typeParsers)
    );
  },

  // :string
  TypeTag: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string(":"),
      r._,
      r.TypeIdentifier,
      (_2, _3, identifier) => {
        return {
          kind: "TypeTag",
          identifier,
        };
      }
    );
  },
});

export function tryParse(str: string): LangType["Program"] {
  const result = Lang.Program.tryParse(str);
  return result;
}

const program = `
export type Point (
  x: number,
  y: number,
)

*.add(a: Point, b: Point) => Point {
  return (
    x: a.x + b.x, 
    y: a.y + b.y
  )
}

.origin() => Point {
  return (x: 0, y: 0)
}

.double Point {
  return (x: arg.x * 2, b: arg.x * 2)
}

.double(a: Point) {
  return (x: a.x * 2, b: b.x * 2)
}

//////////  point.achi
//////////  point.extra.achi // Can extend point.
////////////////// if named differently, error when attempting
////////////////// to define a new function in point

import { Point } from "./point"

Point.unit() => Point {
  return (x: 1, y: 1)
}


const foo = Point(x:3, y:3)

foo = foo
  .double()
  .add(Point(x:3, y:3))



// index.achi

type LinkedMap<T>(
  _map: Map<string, T>,
) as Subbable<T> (
  subscriptors: Set<() => void>,
) as StateHashable (
  mutationHash: number,
)

.set<T>(lm: LinkedMap<T>, key: string, val: T) {
  lm._map.set(key, val);
  Subbable.notify(lm)
  StateHashable.didMutate(lm)
}




// Usage

Point.add(a, b)


const x = Point(x: 1, y: 2)

function add (a: Point, b: Point) {
  
}

const serialize(this: Point) {
  
}

x.serialize()

`;
