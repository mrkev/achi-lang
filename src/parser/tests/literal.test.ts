import { Lang } from "../parser";

test("BooleanLiteral.true", () => {
  const result = Lang.BooleanLiteral.tryParse("true");
  expect(result).toMatchSnapshot();
});

test("BooleanLiteral.false", () => {
  const result = Lang.BooleanLiteral.tryParse("false");
  expect(result).toMatchSnapshot();
});
