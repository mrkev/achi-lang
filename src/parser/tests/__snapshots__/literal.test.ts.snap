// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooleanLiteral.false 1`] = `
{
  "kind": "BooleanLiteral",
  "value": false,
}
`;

exports[`BooleanLiteral.true 1`] = `
{
  "kind": "BooleanLiteral",
  "value": true,
}
`;

exports[`ListLiteral 1`] = `
{
  "expressions": [
    {
      "kind": "NumberLiteral",
      "value": 3,
    },
    {
      "kind": "BooleanLiteral",
      "value": false,
    },
    {
      "kind": "StringLiteral",
      "value": "string",
    },
  ],
  "kind": "ListLiteral",
}
`;

exports[`ListLiteral.recursive 1`] = `
{
  "expressions": [
    {
      "kind": "NumberLiteral",
      "value": 3,
    },
    {
      "expressions": [
        {
          "kind": "BooleanLiteral",
          "value": false,
        },
      ],
      "kind": "ListLiteral",
    },
    {
      "kind": "StringLiteral",
      "value": "string",
    },
  ],
  "kind": "ListLiteral",
}
`;

exports[`MapLiteral 1`] = `
{
  "entries": [
    {
      "expression": {
        "kind": "NumberLiteral",
        "value": 3,
      },
      "identifier": {
        "_meta": {
          "end": {
            "column": 5,
            "line": 1,
            "offset": 4,
          },
          "start": {
            "column": 2,
            "line": 1,
            "offset": 1,
          },
        },
        "kind": "ValueIdentifier",
        "value": "foo",
      },
      "kind": "NamedLiteral",
    },
    {
      "expression": {
        "kind": "BooleanLiteral",
        "value": false,
      },
      "identifier": {
        "_meta": {
          "end": {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
        },
        "kind": "ValueIdentifier",
        "value": "bar",
      },
      "kind": "NamedLiteral",
    },
  ],
  "kind": "MapLiteral",
}
`;

exports[`MapLiteral.recursive 1`] = `
{
  "entries": [
    {
      "expression": {
        "kind": "NumberLiteral",
        "value": 3,
      },
      "identifier": {
        "_meta": {
          "end": {
            "column": 5,
            "line": 1,
            "offset": 4,
          },
          "start": {
            "column": 2,
            "line": 1,
            "offset": 1,
          },
        },
        "kind": "ValueIdentifier",
        "value": "foo",
      },
      "kind": "NamedLiteral",
    },
    {
      "expression": {
        "entries": [
          {
            "expression": {
              "kind": "NumberLiteral",
              "value": 3,
            },
            "identifier": {
              "_meta": {
                "end": {
                  "column": 19,
                  "line": 1,
                  "offset": 18,
                },
                "start": {
                  "column": 16,
                  "line": 1,
                  "offset": 15,
                },
              },
              "kind": "ValueIdentifier",
              "value": "baz",
            },
            "kind": "NamedLiteral",
          },
        ],
        "kind": "MapLiteral",
      },
      "identifier": {
        "_meta": {
          "end": {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
        },
        "kind": "ValueIdentifier",
        "value": "bar",
      },
      "kind": "NamedLiteral",
    },
  ],
  "kind": "MapLiteral",
}
`;
